src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1407,6 +1407,49 @@ void Session::cancel_resumption_delay()
 }
 
 
+void Session::gather_pending_compensating_writes(util::Span<Changeset> changesets,
+                                                 std::vector<ProtocolErrorInfo>* out)
+{
+    if (m_pending_compensating_write_errors.empty()) {
+        return;
+    }
+
+#ifdef REALM_DEBUG
+    REALM_ASSERT_DEBUG(
+        std::is_sorted(m_pending_compensating_write_errors.begin(), m_pending_compensating_write_errors.end(),
+                       [](const ProtocolErrorInfo& lhs, const ProtocolErrorInfo& rhs) {
+                           return lhs.compensating_write_server_version < rhs.compensating_write_server_version;
+                       }));
+#endif
+
+    auto changeset_it = changesets.begin();
+    auto error_it = m_pending_compensating_write_errors.begin();
+    bool found_match = false;
+    while (changeset_it != changesets.end() && error_it != m_pending_compensating_write_errors.end()) {
```

Comment:
If I understood the spec doc correctly, the algorithm should be like this:
```
while (changeset_it != changesets.end() && error_it != m_pending_compensating_write_errors.end()) {
  if (error_it->compensating_write_server_version <= changeset_it->version) {
       found_match = true;
       out->push_back(*error_it);
       ++error_it;
  }
  else {
       ++changeset_it;
  }
}
```

Also, there is a note that the errors should be persisted between sessions until a changeset with a server version greater or equal than the one presented in the error is received. I don't think we do that.

Reply:
^ right, the spec is a bit outdated now. [BAAS-16041](https://jira.mongodb.org/browse/BAAS-16041) changed the way we send compensating writes to send them from history instead of during upload, so the requirement around them persisting across sessions is no longer relevant.

I'll leave a comment on the spec for posterity

Reply:
@kmorkos Does that mean that the error is sent just before the download containing the compensating write? Then I agree that persiting errors is not needed.

Reply:
@jbreams I think my code suggestion still applies though. One optimization would be to use the progress cursor (i.e, download cursor) instead of doing a set intersection.

Reply:
@danieltabacaru it is always sent before the actual compensating write, but not necessarily directly before the download message with the compensating write. e.g. you might have the following sequence
1. ERROR message with warning about compensating write arriving @ V10
2. DOWNLOAD message with V7-9
3. DOWNLOAD message with V10-12 <-- compensating write is a changeset in this download

Reply:
@kmorkos got it. thanks for the explanation. And the way it works I guess is that if we restart the session and v10 has not been acked, we send the error (potentially again).
@jbreams I think we can definitely use the download cursor instead then.

Reply:
I was under the impression that we had to actually find the set intersection of changesets received in download messages and errors received and only report those. Because permissions could change across re-connects, it's possible that you could have a compensating write in server version 11, but only receive v10 and v12. @kmorkos , is that correct or am I mis-remembering?

Reply:
Then I guess you meant `if (changeset_it->version == error_it->compensating_write_server_version)` if you need the exact version. Does that mean that the server can decide not to send history entries at all? Or, will it just send empty changesets? If it's the latter, the we need to check the size as well to make sure we correctly skip the versions.

Reply:
@jbreams the compensating writes are generated & persisted based on your permissions/query at upload time. You are guaranteed to receive a changeset at `compensating_write_server_version` regardless of your permissions/query at the time you receive the download, so @danieltabacaru is right IIUC.

Reply:
That definitely simplifies things then!